### ISSUE TITLE
Asynchronous blob upload with LL API?

### ISSUE THREAD

**Issue by rgacuity** [Azure/azure-iot-sdk-c#562]
- **OS and version used:** <VERSION>  Windows 10 Pro 64-bit

- **SDK version used:** <VERSION> 1.1.3

# Description of the issue:
Is there a function in the LL API corresponding to the high-level API's IoTHubClient_UploadToBlobAsync ?

The docs for IoTHubClient_LL_UploadMultipleBlocksToBlob say it's been deprecated in favor of IoTHubClient_LL_UploadMultipleBlocksToBlob**Async**Ex, but there is no function with that name.

And IoTHubClient_LL_UploadMultipleBlocksToBlobEx blocks until the upload is completed (even though it uses a callback for the actual upload!), i.e. it is not "Async".

**Comment by rgacuity** [comment]
If we need to use the convenience method IoTHubClient_UploadToBlobAsync instead, is there an example that shows how its callback should work?
And when the callback isn't NULL, is it still then necessary to pass in a buffer with all the data to be uploaded in the call to IoTHubClient_UploadToBlobAsync itself?

**Comment by ewertons** [comment]
Hi @rgacuity ,
that's a typo on the comment.
You should use this function: 
```c
IoTHubClient_LL_UploadMultipleBlocksToBlobEx
```

https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/inc/iothub_client_ll.h#L397 

The sample can be found here:
https://github.com/Azure/azure-iot-sdk-c/tree/master/iothub_client/samples/iothub_client_sample_upload_to_blob_mb

**Comment by ewertons** [comment]
We will get that typo fixed right away.

**Comment by rgacuity** [comment]
Thanks, but when I use 
`IoTHubDeviceClient_LL_UploadMultipleBlocksToBlobEx`
I find that that function blocks, i.e. it does not return, until after the entire blob has been uploaded, via multiple calls to the callback.  In other words, that doesn't provide an _**asynchronous**_ upload.

Is there a LL method that does?  

If not, is there an example that shows how to use
`IoTHubClient_UploadToBlobAsync`
with a callback?

**Comment by ewertons** [comment]
Oh, I see.
So this is part of the overall concept of what the LL layer and what the convenience layer are.
The Low Level (LL) layer is non-threaded, blocking.
The upper layer (aka convenience, with no _LL_ on function names) is threaded, async and non-blocking.

So in short, no, there is no LL function that would be non-blocking, you must use the non _LL_ functions.

**Comment by rgacuity** [comment]
OK, so then is there an example that shows how to use
`IoTHubClient_UploadToBlobAsync`
with its callback?

**Comment by yzhong94** [comment]
@ewertons can we put some of the implementation details discussed in our API reference?

**Comment by luzlab** [comment]
Hi @ewertons, I work with @rgacuity and I'm looking at the sample code you provided - thank you btw.

It seems like the return value of `IoTHubDeviceClient_UploadMultipleBlocksToBlobAsync` indicates whether the upload was successful or failed. Is this correct? Does this mean the call to `IoTHubDeviceClient_UploadMultipleBlocksToBlobAsync` blocks until the blob has finished uploading?

```cpp
if (IoTHubDeviceClient_UploadMultipleBlocksToBlobAsync(device_handle, "subdir/hello_world_mb.txt", getDataCallback, NULL) != IOTHUB_CLIENT_OK)
{
  (void)printf("hello world failed to upload\n");
}
else
{
  (void)printf("hello world has been created\n");
}
```

**Comment by rgacuity** [comment]
@ewertons Thanks for the sample.
It looks like we will need to upgrade to a more recent version of the SDK to gain access to the new Device Client functions like `IoTHubDeviceClient_UploadMultipleBlocksToBlobAsync`, as used in your sample.
Until we can try that, @luzlab asks a good question.  The new sample looks very much like the old one for `IoTHubDeviceClient_LL_UploadMultipleBlocksToBlobEx`, which blocks till completion of the upload, in spite of using the same callback to actually upload the data.

**Comment by ewertons** [comment]
Hi @rgacuity and @luzlab ,
why do you say the call to IoTHubDeviceClient_UploadMultipleBlocksToBlobAsync is blocking?

Its return indicates only that the arguments are correct and that the upload thread could be started.
After that all the action happens through the callback function.
Please take a look at the comments here:
https://github.com/Azure/azure-iot-sdk-c/blob/master/iothub_client/inc/iothub_client_core_common.h#L180

**Comment by rgacuity** [comment]
We were just asking, because we'd found that the call to `IoTHubClient_LL_UploadMultipleBlocksToBlobEx`  was blocking, even though it takes the exact same kind of callback as `IoTHubDeviceClient_UploadMultipleBlocksToBlobAsync`.

I haven't yet been able to try the new Device Client functions, because having upgraded to the latest version of the SDK via NuGet, I'm getting a bunch of linker errors (see below).  Any idea what I need to do to fix?

```
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IOTHUB_CLIENT_RESULTStringStorage already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IOTHUB_CLIENT_CONFIRMATION_RESULTStringStorage already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: CALLBACK_TYPEStringStorage already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IOTHUB_CLIENT_RESULTStrings already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IOTHUB_CLIENT_RESULT_FromString already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IOTHUB_CLIENT_CONFIRMATION_RESULTStrings already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IOTHUB_CLIENT_CONFIRMATION_RESULT_FromString already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_CreateFromConnectionString already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_Create already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_CreateWithTransport already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_Destroy already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SendEventAsync already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_GetSendStatus already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SetMessageCallback already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SetConnectionStatusCallback already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SetRetryPolicy already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_GetRetryPolicy already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_GetLastMessageReceiveTime already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_DoWork already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SetOption already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SetDeviceTwinCallback already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SendReportedState already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SetDeviceMethodCallback already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SetDeviceMethodCallback_Ex already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_DeviceMethodResponse already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_UploadToBlob already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_UploadMultipleBlocksToBlob already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_UploadMultipleBlocksToBlobEx already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: CALLBACK_TYPEStrings already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: CALLBACK_TYPE_FromString already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_CreateFromDeviceAuth already defined in iothub_client.lib(iothub_client_ll.obj)
...
1>iothub_client.lib(iothub_client_core_ll.obj) : error LNK2019: unresolved external symbol IoTHubClient_Auth_Get_TrustBundle referenced in function IoTHubClientCore_LL_CreateFromEnvironment
1>iothub_client.lib(iothub_client_edge.obj) : error LNK2001: unresolved external symbol IoTHubClient_Auth_Get_TrustBundle
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IOTHUB_CLIENT_RESULTStringStorage already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IOTHUB_CLIENT_CONFIRMATION_RESULTStringStorage already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: CALLBACK_TYPEStringStorage already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IOTHUB_CLIENT_RESULTStrings already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IOTHUB_CLIENT_RESULT_FromString already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IOTHUB_CLIENT_CONFIRMATION_RESULTStrings already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IOTHUB_CLIENT_CONFIRMATION_RESULT_FromString already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_CreateFromConnectionString already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_Create already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_CreateWithTransport already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_Destroy already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SendEventAsync already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_GetSendStatus already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SetMessageCallback already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SetConnectionStatusCallback already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SetRetryPolicy already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_GetRetryPolicy already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_GetLastMessageReceiveTime already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_DoWork already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SetOption already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SetDeviceTwinCallback already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SendReportedState already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SetDeviceMethodCallback already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_SetDeviceMethodCallback_Ex already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_DeviceMethodResponse already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_UploadToBlob already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_UploadMultipleBlocksToBlob already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_UploadMultipleBlocksToBlobEx already defined in iothub_client.lib(iothub_client_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: CALLBACK_TYPEStrings already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: CALLBACK_TYPE_FromString already defined in iothub_client.lib(iothub_client_core_ll.obj)
1>iothub_client_amqp_transport.lib(iothub_client_ll.obj) : error LNK2005: IoTHubClient_LL_CreateFromDeviceAuth already defined in iothub_client.lib(iothub_client_ll.obj)
1>   Creating library C:\projects\cli\x64\Debug\CrossCutting.lib and object C:\projects\cli\x64\Debug\CrossCutting.exp
1>AzureIoTDevice.obj : error LNK2001: unresolved external symbol "public: virtual struct QMetaObject const * __cdecl AzureIoTDevice::metaObject(void)const " (?metaObject@AzureIoTDevice@@UEBAPEBUQMetaObject@@XZ)
1>AzureIoTDevice.obj : error LNK2001: unresolved external symbol "public: virtual void * __cdecl AzureIoTDevice::qt_metacast(char const *)" (?qt_metacast@AzureIoTDevice@@UEAAPEAXPEBD@Z)
1>AzureIoTDevice.obj : error LNK2001: unresolved external symbol "public: virtual int __cdecl AzureIoTDevice::qt_metacall(enum QMetaObject::Call,int,void * *)" (?qt_metacall@AzureIoTDevice@@UEAAHW4Call@QMetaObject@@HPEAPEAX@Z)
1>AzureIoTDevice.obj : error LNK2001: unresolved external symbol "public: static struct QMetaObject const AzureIoTDevice::staticMetaObject" (?staticMetaObject@AzureIoTDevice@@2UQMetaObject@@B)
1>iothub_client.lib(iothub_client_core_ll.obj) : error LNK2019: unresolved external symbol IoTHubClient_Auth_Get_TrustBundle referenced in function IoTHubClientCore_LL_CreateFromEnvironment
1>iothub_client.lib(iothub_client_edge.obj) : error LNK2001: unresolved external symbol IoTHubClient_Auth_Get_TrustBundle
```

**Comment by rgacuity** [comment]
It looks like the problem happens when trying to include both iothub_device_client.h and iothub_client.h in a single source file.  Is there an incompatibility between the old client and new device client APIs?

**Comment by ewertons** [comment]
@rgacuity , yes you figured out the name collision and how to fix it. Good job.
So initially we had two C modules, one for the device client (which talks to the IoT Hub) and one for the service client (which talks to the Event hub behind the IoT Hub).
The names of the modules were:
iothub_client.h
iothub_service_client.h

So we are renaming things now and iothub_client.h will become iothub_device_client.h for consistency.
However, since we hadn't  bumped the major version of the SDK we cannot simply delete iothub_client.h, so both need to coexist for a while.

We apologize for the confusion. 
We know it would have been cleaner bumping the major version, but we are still working on other improvements until we get to that point.

So long story short, you can stop using iothub_client.h and start using iothub_device_client.h.

**Comment by rgacuity** [comment]
OK, thanks for the clarification.  As it is, I believe I've gotten this working with `IoTHubClient_UploadMultipleBlocksToBlobAsyncEx`, such that I don't need to upgrade the SDK, nor change over to using iothub_device_client.h.

**Comment by ewertons** [comment]
Thanks @rgacuity and @luzlab . Let us know if we can help in any other way

**Comment by az-iot-builder-01** [comment]
@rgacuity, @rgacuity, @rgacuity, @rgacuity, @luzlab, @rgacuity, @rgacuity, @rgacuity, @rgacuity, @rgacuity, thank you for your contribution to our open-sourced project! Please help us improve by filling out this 2-minute [customer satisfaction survey](https://www.surveymonkey.com/r/Q5PLZ3V)

**Comment by dfabianZF** [comment]
Hello,

I came across this issue and I would like to add my point of view. I use the SDK in a single-threaded C++ application that uses boost::asio. I only use the LL API and do my own scheduling by invoking `IoTHubDeviceClient_LL_DoWork()` in a timer.

>Oh, I see.
So this is part of the overall concept of what the LL layer and what the convenience layer are.
The Low Level (LL) layer is non-threaded, blocking.

Actually, this is not entirely true. The LL API contain asynchronous (non-blocking) event messaging. I can do a lot of non-blocking code within a single thread and send messages to the hub without any trouble using `IoTHubDeviceClient_LL_SendEventAsync()`. This function does not block and allows me to wait for the result using a callback. I expected a similar behavior from the blob upload function but as stated above, it does not work that way.

The proposed solution here is to use the convenience API but then I have a mixed code with LL and non-LL functions and most importantly I already have an LL connection handle which I cannot use in non-LL functions. There really should be something like `IoTHubClient_LL_UploadMultipleBlocksToBlobAsyncEx()` which retains the original declaration but calls the callback asynchronously when a chunk has been sent.

**Comment by jspaith** [comment]
@dfabianZF - could you please open a new issue to track this, referencing back to this?  We don't really track closed issues.  Thanks

**Comment by dfabianZF** [comment]
Here you go #1091.
